Mathlib/CategoryTheory/Triangulated/Subcategory.lean

```diff
@@ -40,111 +42,141 @@ assert_not_exists TwoSidedIdeal
 
 namespace CategoryTheory
 
-open Category Limits Preadditive ZeroObject
+open Category Limits Preadditive ZeroObject Pretriangulated Triangulated
 
-namespace Triangulated
+variable {C : Type*} [Category C] [HasZeroObject C] [HasShift C ℤ]
+  [Preadditive C] [∀ (n : ℤ), (shiftFunctor C n).Additive] [Pretriangulated C]
 
-open Pretriangulated
+namespace ObjectProperty
 
-variable (C : Type*) [Category C] [HasZeroObject C] [HasShift C ℤ]
-  [Preadditive C] [∀ (n : ℤ), (shiftFunctor C n).Additive] [Pretriangulated C]
+variable (P : ObjectProperty C)
 
-/-- A triangulated subcategory of a pretriangulated category `C` consists of
-a predicate `P : C → Prop` which contains a zero object, is stable by shifts, and such that
-if `X₁ ⟶ X₂ ⟶ X₃ ⟶ X₁⟦1⟧` is a distinguished triangle such that if `X₁` and `X₃` satisfy
-`P` then `X₂` is isomorphic to an object satisfying `P`. -/
-structure Subcategory where
-  /-- the underlying predicate on objects of a triangulated subcategory -/
-  P : ObjectProperty C
-  zero' : ∃ (Z : C) (_ : IsZero Z), P Z
-  shift (X : C) (n : ℤ) : P X → P (X⟦n⟧)
-  ext₂' (T : Triangle C) (_ : T ∈ distTriang C) : P T.obj₁ → P T.obj₃ → P.isoClosure T.obj₂
+/-- Given `P : ObjectProperty C` with `C` a pretriangulated category, this
+is the property that whenever `X₁ ⟶ X₂ ⟶ X₃ ⟶ X₁⟦1⟧` is a distinguished triangle
+such that `P X₂` and `P X₃` hold, then `P.isoClosure X₁` holds. -/
+class IsTriangulatedClosed₁ : Prop where
+  ext₁' (T : Triangle C) (_ : T ∈ distTriang C) : P T.obj₂ → P T.obj₃ → P.isoClosure T.obj₁
 
-namespace Subcategory
+/-- Given `P : ObjectProperty C` with `C` a pretriangulated category, this
+is the property that whenever `X₁ ⟶ X₂ ⟶ X₃ ⟶ X₁⟦1⟧` is a distinguished triangle
+such that `P X₁` and `P X₃` hold, then `P.isoClosure X₂` holds. -/
+class IsTriangulatedClosed₂ : Prop where
+  ext₂' (T : Triangle C) (_ : T ∈ distTriang C) : P T.obj₁ → P T.obj₃ → P.isoClosure T.obj₂
 
-variable {C}
-variable (S : Subcategory C)
-
-lemma zero [S.P.IsClosedUnderIsomorphisms] : S.P 0 := by
-  obtain ⟨X, hX, mem⟩ := S.zero'
-  exact S.P.prop_of_iso hX.isoZero mem
-
-/-- The closure under isomorphisms of a triangulated subcategory. -/
-def isoClosure : Subcategory C where
-  P := S.P.isoClosure
-  zero' := by
-    obtain ⟨Z, hZ, hZ'⟩ := S.zero'
-    exact ⟨Z, hZ, Z, hZ', ⟨Iso.refl _⟩⟩
-  shift X n := by
-    rintro ⟨Y, hY, ⟨e⟩⟩
-    exact ⟨Y⟦n⟧, S.shift Y n hY, ⟨(shiftFunctor C n).mapIso e⟩⟩
+/-- Given `P : ObjectProperty C` with `C` a pretriangulated category, this
+is the property that whenever `X₁ ⟶ X₂ ⟶ X₃ ⟶ X₁⟦1⟧` is a distinguished triangle
+such that `P X₁` and `P X₂` hold, then `P.isoClosure X₃` holds. -/
+class IsTriangulatedClosed₃ : Prop where
+  ext₃' (T : Triangle C) (_ : T ∈ distTriang C) : P T.obj₁ → P T.obj₂ → P.isoClosure T.obj₃
+
+lemma ext_of_isTriangulatedClosed₁'
```

Comment:
I had understood that mathlib sought to avoid primed-lemmas, and that in any case, they should be accompanied by a docstring that explains their relation with the corresponding non-primed lemma.